Add interpolation option for unwrapping

mirzaees opened the pull request:
- Added interpolation function for masked pixels before unwrapping to ease the computations
- Added unit test and changed sample interferograms to a simple ramp instead of all zero

hfattahi commented on src/dolphin/unwrap/_unwrap.py:
```diff
@@ -304,20 +324,52 @@ def unwrap(
+
+        ifg = io.load_gdal(ifg_filename)
+        corr = io.load_gdal(corr_filename)
+        logger.info("Masking pixels with correlation above 0.7")
+        coherent_pixel_mask = corr[:] >= 0.7
```
I don't think if it should be the responsibility of _unwrap.py to define a mask. I think the mask should be defined upstream and passed to unwrap.

hfattahi commented on src/dolphin/interpolation.py:
```diff
@@ -0,0 +1,149 @@
+
+
+@numba.njit(parallel=True)
+def _interp_loop(
+    ifg, weights, weight_cutoff, num_neighbors, alpha, indices, interpolated_ifg
+):
```
I wonder if you could add a simple unit test for this function. for example if we assume a simple phase ramp with a known model coefficients and randomly mask out pixels as invalid pixels, and interpolate the phase of valid pixels, then I expect the interpolated phase over invalid pixels to be close to the value we get from the model coefficients with a tolerance.

hfattahi replied:
@mirzaees I don't see this unit test added? Are you planning to add or have a different plan?

hfattahi commented on src/dolphin/interpolation.py:
```diff
@@ -0,0 +1,149 @@
+                        break
+
+            # TODO : why use the "counter - 1" here to normalize?
+            r2_norm = (r2[counter - 1] ** alpha) / 2
```
It looks like the intrinsic assumption here is that the last element of r2 is the largest element. Is that correct assumption? if the largest element is needed why not just get the max of r2? 
I think I do not understans the arbitrary division by 2 here either. I don't know if you or @scottstanie or @taliboliver can help me understand what is going on here.

taliboliver replied:
Hi @hfattahi yes I think using max of r^2 should be equivalent and about the division by two my understanding was that is part of the scaling factor from Ke Wang implementation (https://github.com/UT-Radar-Interferometry-Group/psinterp/blob/91b5e797f7768766dfb9ab941ba47cdba479472c/src/cpp/psinterp.cpp#L117) I will take a deeper look into it and add more info on this.

scottstanie replied:
![image](https://github.com/isce-framework/dolphin/assets/8291800/382a2dcb-2a93-491c-a8e1-ef83bc0eada6)
1. We could add a reference like `See Eq. (2), Chen 2015`


2. Thanks for asking about the r2, which I tried to explain more [here in a comment](https://github.com/mirzaees/dolphin/pull/1/files#diff-4802b88e403e32ee9c9c7dc279fd6dea658ecbb460a57149addc299b9826a0d3R129-R130) 
But your question made me realize that I did make a change from Ke's code in generating the circle indexes:
https://github.com/mirzaees/dolphin/blob/7115a57e8861766b04f949eb0bd8fa1ceaa02267/src/dolphin/similarity.py#L241-L243

If we know we are going to access all the pixels in the circle
![image](https://github.com/isce-framework/dolphin/assets/8291800/bfcf947e-3c27-4aec-96f7-6d234c47ab74)
then sorting them leads to faster runtime. 

BUT that breaks this assumption here, as his code produced the indexes in a spiral outward so the last one was always at the max radius $R$. So we should probably change the code to `get_circle_idxs(... sort_output=False)` for this function

hfattahi replied:
Thanks for the clarification. Let me know if the "get_circle_idxs" is already fixed so that we resolve this comment.

mirzaees replied:
Hi @hfattahi, yes it is fixed now. We applied all you suggestions, please take another look so we can merge

hfattahi reviewed (commented):
Thanks @mirzaees for this useful addition. I have few comments below. Also if I understand correctly, it looks like the current implementation allows to either use run_goldstein or run_interpolation or neither of them. I don't see a functionality to run both. Is that correct?

scottstanie commented on src/dolphin/interpolation.py:
```diff
@@ -0,0 +1,149 @@
+    alpha: float = 0.75,
+    weight_cutoff: float = 0.0,
+) -> np.ndarray:
+    r"""Interpolation of persistent scatterers.
```
My PR here https://github.com/mirzaees/dolphin/pull/1#issuecomment-1994571567 has a further expansion on the docstring

hfattahi reviewed (commented):
Thanks for addressing my comments. My final comment is that the current unit test does not really check if the interpolation function does the right thing or not. Would it be possible to add my suggested test below?

mirzaees commented:
> Thanks for addressing my comments. My final comment is that the current unit test does not really check if the interpolation function does the right thing or not. Would it be possible to add my suggested test below?

Thanks @hfattahi for your suggestion, I added a testing function for that. Also changed the sample interferograms a bit to be closer to reality. the sample interferograms were all zero

hfattahi commented on tests/test_unwrap.py:
```diff
@@ -119,28 +122,59 @@ def test_unwrap_snaphu_nodata(self, tmp_path, list_of_gtiff_ifgs, corr_raster):
+        )
+        assert math.isclose(
+            np.angle(interpolated_ifg[70, 101]), 0.56214905, rel_tol=0.0000001
+        )
```
Glad to see some numerical check of the interpolation function. Would you please explain how do you expect the value to be 0.73608005 and 0.56214905 at those two selected pixels after interpolation?

mirzaees replied:
I ran the function locally and checked the results, since the interferogram has some fixed real values, the interpolation should be the same.

hfattahi replied:
If I understand correctly the test is working with an interferogram that has a constant phase (same phase at all pixels). So you expect the interpolation will lead to the same constant phase. Is that correct?

mirzaees replied:
The sample interferogram looks like the attached figure, and I am interpolating for the middle masked line, I expect the interpolation would always lead to the same phase values

<img width="415" alt="Screenshot 2024-03-18 at 3 15 25 PM" src="https://github.com/isce-framework/dolphin/assets/37273875/daca18d3-a007-4ed2-8a5c-3d87241ad6f1">

hfattahi replied:
Thanks @mirzaees for the plot. Can you elaborate what do you mean by the "Same phase value"?

hfattahi reviewed (commented):
Please see below for my suggestion to make the test more clear. Please feel free to change the detailed implementation. I just meant to give a sudo code to describe the idea. In general please be more generous in commenting the code so that we can always get back to them and understand what we have done or what were our assumptions/algorithms etc. We are very close. Thanks for your patience and addressing my comments.

mirzaees commented on tests/test_unwrap.py:
```diff
@@ -119,13 +122,59 @@ def test_unwrap_snaphu_nodata(self, tmp_path, list_of_gtiff_ifgs, corr_raster):
+        ) / 8
+        assert math.isclose(
+            np.angle(interpolated_ifg[70, 100]), neighbors_average_value, rel_tol=1e-02
+        )
```
Thanks @hfattahi, this is absolutely more transparent

hfattahi reviewed (approved):
LGTM. Thanks @mirzaees and thanks to @taliboliver and @scottstanie for their initial work on this.
